table_cleanup.go

```diff
@@ -0,0 +1,97 @@
+package kbucket
+
+import (
+	"context"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+func (rt *RoutingTable) cleanup() {
+	validatePeerF := func(p peer.ID) bool {
+		queryCtx, cancel := context.WithTimeout(rt.ctx, rt.peerValidationTimeout)
+		defer cancel()
+		return rt.peerValidationFnc(queryCtx, p)
+	}
+
+	cleanupTickr := time.NewTicker(rt.tableCleanupInterval)
+	defer cleanupTickr.Stop()
+	for {
+		select {
+		case <-cleanupTickr.C:
+			ps := rt.peersToValidate()
+			for _, pinfo := range ps {
+				// TODO This is racy
+				// A peer could disconnect immediately after we validate it & would thus be in missing state again
+				// which means we would wrongly mark it as active here. The longer term solution is to
+				// handle all peer related events in a single event loop in the RT or more fingrained locking at bucket/peer level.
+				// See https://github.com/libp2p/go-libp2p-kbucket/issues/60
+				if validatePeerF(pinfo.Id) {
+					rt.tabLock.Lock()
+					// add it back/mark it as active ONLY if it is still in the RT
+					// to avoid adding it back if it's been marked as dead
+					i := rt.bucketIdForPeer(pinfo.Id)
+					if rt.buckets[i].getPeer(pinfo.Id) != nil {
+						log.Infof("successfully validated missing peer=%s, marking it as active", pinfo.Id)
+						rt.addPeer(pinfo.Id)
```

Comment:
Can't we just set the peer state to active here instead of calling `addPeer`?

Reply:
Done.